### PR TITLE
ISSUE-436 Respect contract described in StorageManager for JdbcStorageManager

### LINKED ISSUE #436: JdbcStorageManager doesn't respect the contract in StorageManager
The contract in StorageManager is a bit tough since it is designed based on in-memory first which is easy to achieve this without worrying about overhead.

I found that JdbcStorageManager doesn't respect the contract in StorageManager: some operation is already respecting it but without transactional manner. It may hurt overall performance on usage if we make JdbcStorageManager implementing StorageManager correctly, but looks like both registry and streamline are relying on such contract, so might be better to ensure once, and look for alternative approaches.

### PR DESCRIPTION
**Pull request by HeartSaVioR**
* ensure that JdbcStorageManager implements StorageManager correctly with transactional manner
* this assumes we're aware about the impact on performance

This closes #436

### CODE REVIEW

File: storage/core/src/main/java/com/hortonworks/registries/storage/impl/jdbc/JdbcStorageManager.java [hunk 1]
```diff
@@ -61,29 +62,96 @@ public JdbcStorageManager(QueryExecutor queryExecutor) {
 
     @Override
     public void add(Storable storable) throws AlreadyExistsException {
-        log.debug("Adding storable [{}]", storable);
-        queryExecutor.insert(storable);
+        boolean committed = false;
+        try {
+            beginTransaction(TransactionIsolation.DEFAULT);
```
Comment (arunmahadevan) [inline storage/core/src/main/java/com/hortonworks/registries/storage/impl/jdbc/JdbcStorageManager.java]: the transactions are handled at a higher level via [ManagedTransaction](https://github.com/hortonworks/registry/blob/master/storage/core/src/main/java/com/hortonworks/registries/storage/transaction/ManagedTransaction.java).

I dont think the Storage Manager should support txn for individual operations. There could be multiple storage manager operations that needs to be part of a txn.

**Comment by HeartSaVioR** [comment]
@arunmahadevan 
Nested transaction is supported now, so that's not impossible to use transaction in each operation. @raju-saravanan Could you confirm?
And there's currently no way to get current transaction, so can't conditionally wrap the logic with transaction. The implementation of JdbcStorageManager should not depend on transaction outside, and such requirement should be documented if it should.

Now the storage module is a common library for three projects, and this is not the first time I claim about broken contract. so I would want to do the thing the right way.

cc. @satishd

**Comment by raju-saravanan** [comment]
@HeartSaVioR and @arunmahadevan Yes, we support nested transactions. If a transaction is nested then it uses the connection details of the outer transaction, it commits or rollbacks with the outermost transaction block.

File: storage/core/src/main/java/com/hortonworks/registries/storage/impl/jdbc/JdbcStorageManager.java [hunk 2]
```diff
@@ -61,29 +62,96 @@ public JdbcStorageManager(QueryExecutor queryExecutor) {
...
     public <T extends Storable> T remove(StorableKey key) throws StorageException {
-        T oldVal = get(key);
-        if (key != null) {
-            log.debug("Removing storable key [{}]", key);
-            queryExecutor.delete(key);
+        boolean committed = false;
+        try {
+            beginTransaction(TransactionIsolation.DEFAULT);
+
+            T oldVal = get(key);
```
Comment (raju-saravanan) [inline storage/core/src/main/java/com/hortonworks/registries/storage/impl/jdbc/JdbcStorageManager.java]: get() here may not be required. As by default "delete" command with "where" clause which doesn't match any records when executed doesn't throw any error, but quietly prints that "0 rows deleted". I confirmed this behavior in MySQL, Postgres and Oracle.

File: storage/core/src/main/java/com/hortonworks/registries/storage/exception/NotFoundException.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.hortonworks.registries.storage.exception;
+
+public class NotFoundException extends StorageException {
```
Comment (raju-saravanan) [inline storage/core/src/main/java/com/hortonworks/registries/storage/exception/NotFoundException.java]: nit: There are couple of "[a-zA-Z]*NotFoundException", maybe we can call this "StorableNotFoundException" to distinguish them from this exception :)

Comment (HeartSaVioR) [reply storage/core/src/main/java/com/hortonworks/registries/storage/exception/NotFoundException.java]: Nice finding. Will fix.

File: storage/core/src/main/java/com/hortonworks/registries/storage/impl/jdbc/JdbcStorageManager.java (hunk 2 above)
Comment (HeartSaVioR) [reply storage/core/src/main/java/com/hortonworks/registries/storage/impl/jdbc/JdbcStorageManager.java]: The method needs to return old value as method contract / signature, so need to get old value first.

Comment (raju-saravanan) [reply storage/core/src/main/java/com/hortonworks/registries/storage/impl/jdbc/JdbcStorageManager.java]: I see that there are instances were storage entity obtained by calling remove() is being used and certain places where we just discard the value. I think we should have remove() not return any values and if the user wants to fetch the entity which is being removed then he can do get() and followed by remove(). I know this is an existing behavior not related to this PR, if you agree with the proposal we can come back to it in a different PR.

**Comment by HeartSaVioR** [comment]
Thanks @raju-saravanan for the reviewing. I'll wait for @arunmahadevan and @satishd to review this, and merge once either one approves the patch.

**Comment by HeartSaVioR** [comment]
@satishd @arunmahadevan  
As I said earlier, I'm not sure it is safe to rely on Postgre's error code, so don't want to do it on my own. If anyone would like to do it, please feel free to take the issue up.

If we can't make sure the implementation respects the interface, I would like to keep others as it is, and modify only addOrUpdate to do get -> add/update guarding with transaction. It would be easier to make a change to JdbcStorageManager, but I'm OK to give a try to make a change in PostgresqlExecutor. Does it make sense then?

**Comment by satishd** [comment]
@HeartSaVioR
>but I'm OK to give a try to make a change in PostgresqlExecutor. Does it make sense then?
I agree on making a change in `PostgresqlExecutor#insertOrUpdate`.

**Comment by HeartSaVioR** [comment]
>> but I'm OK to give a try to make a change in PostgresqlExecutor. Does it make sense then?

> I agree on making a change in PostgresqlExecutor#insertOrUpdate.

OK I can handle that. Will address from #441.

**Pull request closed without merging**